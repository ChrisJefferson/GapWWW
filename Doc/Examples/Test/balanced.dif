1,7d0
<   if P is not perfect
<   then find next w(x,y)
<   else if Index(P,&lang;sgens&rang;) does not complete or Index(P,&lang;sgens&rang;) &ne; n
<   then find next w(x,y)
<   else if Index(P,&lang;y&rang;) does not complete or Index(P,&lang;y&rang;) &ne; Size(G)/p
<   then find next w(x,y)
<   else a suitable w(x,y) has been found.
129c122,208
<                [ ... 87 lines deleted ... ]
---
>  #bisyllables = 25 (#bisyllables in tail = 10) #words tested: 2
> ACEStats:
>   index=9 cputime=0 ms maxcosets=569 totcosets=588
> Large subgroup index OK
> ACEStats for cyclic subgroup:
>   index=168 cputime=280 ms maxcosets=115278 totcosets=118488
> Cyclic subgroup index OK
> Success! ... new relator: 
>    x*y*x*y*x*y^-1*x*y^-1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^
> -1*x*y*x*y*x*y^-1*x*y*x*y^-1*x*y*x*y*x*y^-1*x*y^-1
> ... continuing (there may be a shorter relator).
> Candidate relator: x*y*x*y*x*y^-1*x*y^-1*x*y^
> -1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^-1*x*y^-1*x*y*x*y*x*y*x*y^
> -1*x*y^-1*x*y^-1
>  #bisyllables = 23 (#bisyllables in tail = 8) #words tested: 4
> ACEStats:
>   index=9 cputime=0 ms maxcosets=205 totcosets=253
> Large subgroup index OK
> ACEStats for cyclic subgroup:
>   index=168 cputime=0 ms maxcosets=370 totcosets=472
> Cyclic subgroup index OK
> Success! ... new relator: 
>    x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^
> -1*x*y^-1*x*y*x*y*x*y*x*y^-1*x*y^-1*x*y^-1
> ... continuing (there may be a shorter relator).
> Candidate relator: x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^-1*x*y^
> -1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^-1*x*y^-1
>  #bisyllables = 21 (#bisyllables in tail = 6) #words tested: 14
> ACEStats:
>   index=9 cputime=0 ms maxcosets=327 totcosets=336
> Large subgroup index OK
> ACEStats for cyclic subgroup:
>   index=168 cputime=30 ms maxcosets=18527 totcosets=18675
> Cyclic subgroup index OK
> Success! ... new relator: 
>    x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^-1*x*y^
> -1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^-1*x*y^-1
> ... continuing (there may be a shorter relator).
> Candidate relator: x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^
> -1*x*y^-1*x*y*x*y*x*y*x*y*x*y*x*y^-1*x*y*x*y*x*y
>  #bisyllables = 19 (#bisyllables in tail = 4) #words tested: 128
> ACEStats:
>   index=9 cputime=0 ms maxcosets=181 totcosets=193
> Large subgroup index OK
> ACEStats for cyclic subgroup:
>   index=0 cputime=1290 ms maxcosets=666438 totcosets=666664
> Candidate relator: x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^
> -1*x*y*x*y^-1*x*y*x*y*x*y*x*y*x*y*x*y^-1*x*y*x*y
>  #bisyllables = 19 (#bisyllables in tail = 4) #words tested: 192
> ACEStats:
>   index=9 cputime=0 ms maxcosets=185 totcosets=203
> Large subgroup index OK
> ACEStats for cyclic subgroup:
>   index=0 cputime=1340 ms maxcosets=666474 totcosets=666664
> Candidate relator: x*y*x*y*x*y^-1*x*y*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^
> -1*x*y^-1*x*y*x*y*x*y*x*y*x*y^-1*x*y*x*y*x*y
>  #bisyllables = 19 (#bisyllables in tail = 4) #words tested: 255
> ACEStats:
>   index=9 cputime=0 ms maxcosets=205 totcosets=215
> Large subgroup index OK
> ACEStats for cyclic subgroup:
>   index=0 cputime=1350 ms maxcosets=666476 totcosets=666664
> Candidate relator: x*y*x*y*x*y^-1*x*y*x*y^-1*x*y*x*y*x*y^-1*x*y*x*y*x*y*x*y^
> -1*x*y*x*y*x*y*x*y^-1*x*y*x*y^-1*x*y^-1
>  #bisyllables = 19 (#bisyllables in tail = 4) #words tested: 275
> ACEStats:
>   index=9 cputime=0 ms maxcosets=185 totcosets=249
> Large subgroup index OK
> ACEStats for cyclic subgroup:
>   index=168 cputime=0 ms maxcosets=488 totcosets=511
> Cyclic subgroup index OK
> Success! ... new relator: 
>    x*y*x*y*x*y^-1*x*y*x*y^-1*x*y*x*y*x*y^-1*x*y*x*y*x*y*x*y^
> -1*x*y*x*y*x*y*x*y^-1*x*y*x*y^-1*x*y^-1
> ... continuing (there may be a shorter relator).
> Candidate relator: x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y^-1*x*y^-1*x*y^
> -1*x*y*x*y*x*y^-1*x*y*x*y*x*y*x*y^-1
>  #bisyllables = 17 (#bisyllables in tail = 2) #words tested: 628
> ACEStats:
>   index=9 cputime=0 ms maxcosets=173 totcosets=177
> Large subgroup index OK
> ACEStats for cyclic subgroup:
>   index=168 cputime=0 ms maxcosets=616 totcosets=637
> Cyclic subgroup index OK
> Success! ... new relator: 
>    x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y^
> -1*x*y*x*y*x*y*x*y^-1
