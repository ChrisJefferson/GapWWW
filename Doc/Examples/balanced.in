F := FreeGroup("x","y");;
x := F.1;; y := F.2;;
p := 5;
nexpts := p-1;
expts := Concatenation([1 .. Int(p/2)], 
[-Int(p/2) .. -1]);
sylls := List(expts, i -> x*y^i);
cLength := 6;
c := ListWithIdenticalEntries(cLength, expts[1]);
word := vec -> Product(vec, veci -> x*y^expts[veci], One(x));
word(c);
getNext := function(vec)
local i;
   i := 1;
   while i <= cLength and vec[i] mod p = p-1 do
      vec[i] := 1;
      i := i + 1;
   od;
   if i <= cLength then
      vec[i] := vec[i] + 1;
   else
      vec := fail;
   fi;
end;
printVecWord := function(vec, num)
local i;
   Print(vec, "  ", word(vec), "\n");
   i := 1;;
   while i < num and vec <> fail do
   i := i + 1;
   getNext(vec);
   Print(vec, "  ", word(vec), "\n");
   od;
end;
printVecWord( [1,1,1,1,1,1], 20);
LoadPackage("ace");
SetInfoLevel(InfoACE,0);
ReadPackage("ace","res-examples/pgrelfind.g");
SetInfoLevel(InfoACE,1);
f:= FreeGroup("x","y");;
x:= f.1;; y:= f.2;;
rels:=[ x^2, y^3,
y^(-1)*x*y^(-1)*x*y^(-1)*x*y^(-1)*x*y^(-1)*x*y^(-1)*x*y^(-1)*x,
x*y^(-1)*x*y*x*y^(-1)*x*y*x*y^(-1)*x*y*x*y^(-1)*x*y*x*y*x*
y^(-1)*x*y*x*y^(-1)*x*y*x*y^(-1)*x*y*x*y^(-1)*x*y*x*y ];;
G:=f/rels;
Size(G);
l:=LowIndexSubgroupsFpGroup( G, TrivialSubgroup( G),10);
sgens:=GeneratorsOfGroup(l[2]);
sgens:=List(sgens,UnderlyingElement);
fgens := [ x, y ];
PGRelFind(fgens, rels, sgens );
newrels := [ x^2, y^3, x*y*x*y*x*y^(-1)*x*y^(-1)*x*y^(-1)*x*y*x*
y*x*y^(-1)*x*y^(-1)*x*y*x*y^(-1)*x*y^(-1)*x*y*x*y^(-1)*x*y ];
ge:=f/newrels;
Size(ge);
ExponentSumWord(newrels[3],x);
ExponentSumWord(newrels[3],y);
balancedrels := [x^2*y^3, newrels[3]*x^-16];
gb := f/balancedrels;
Size(gb);
quit;
