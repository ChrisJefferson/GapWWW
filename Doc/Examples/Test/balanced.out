  if P is not perfect
  then find next w(x,y)
  else if Index(P,&lang;sgens&rang;) does not complete or Index(P,&lang;sgens&rang;) &ne; n
  then find next w(x,y)
  else if Index(P,&lang;y&rang;) does not complete or Index(P,&lang;y&rang;) &ne; Size(G)/p
  then find next w(x,y)
  else a suitable w(x,y) has been found.
gap> F := FreeGroup("x","y");;
gap> x := F.1;; y := F.2;;
gap> p := 5;
5
gap> nexpts := p-1;
4
gap> expts := Concatenation([1 .. Int(p/2)], 
> [-Int(p/2) .. -1]);
[ 1, 2, -2, -1 ]
gap> sylls := List(expts, i -> x*y^i);
[ x*y, x*y^2, x*y^-2, x*y^-1 ]
gap> cLength := 6;
6
gap> c := ListWithIdenticalEntries(cLength, expts[1]);
[ 1, 1, 1, 1, 1, 1 ]
gap> word := vec -> Product(vec, veci -> x*y^expts[veci], One(x));
function( vec ) ... end
gap> word(c);
x*y*x*y*x*y*x*y*x*y*x*y
gap> getNext := function(vec)
> local i;
>    i := 1;
>    while i <= cLength and vec[i] mod p = p-1 do
>       vec[i] := 1;
>       i := i + 1;
>    od;
>    if i <= cLength then
>       vec[i] := vec[i] + 1;
>    else
>       vec := fail;
>    fi;
> end;
function( vec ) ... end
gap> printVecWord := function(vec, num)
> local i;
>    Print(vec, "  ", word(vec), "\n");
>    i := 1;;
>    while i < num and vec <> fail do
>    i := i + 1;
>    getNext(vec);
>    Print(vec, "  ", word(vec), "\n");
>    od;
> end;
function( vec, num ) ... end
gap> printVecWord( [1,1,1,1,1,1], 20);
[ 1, 1, 1, 1, 1, 1 ]  x*y*x*y*x*y*x*y*x*y*x*y
[ 2, 1, 1, 1, 1, 1 ]  x*y^2*x*y*x*y*x*y*x*y*x*y
[ 3, 1, 1, 1, 1, 1 ]  x*y^-2*x*y*x*y*x*y*x*y*x*y
[ 4, 1, 1, 1, 1, 1 ]  x*y^-1*x*y*x*y*x*y*x*y*x*y
[ 1, 2, 1, 1, 1, 1 ]  x*y*x*y^2*x*y*x*y*x*y*x*y
[ 2, 2, 1, 1, 1, 1 ]  x*y^2*x*y^2*x*y*x*y*x*y*x*y
[ 3, 2, 1, 1, 1, 1 ]  x*y^-2*x*y^2*x*y*x*y*x*y*x*y
[ 4, 2, 1, 1, 1, 1 ]  x*y^-1*x*y^2*x*y*x*y*x*y*x*y
[ 1, 3, 1, 1, 1, 1 ]  x*y*x*y^-2*x*y*x*y*x*y*x*y
[ 2, 3, 1, 1, 1, 1 ]  x*y^2*x*y^-2*x*y*x*y*x*y*x*y
[ 3, 3, 1, 1, 1, 1 ]  x*y^-2*x*y^-2*x*y*x*y*x*y*x*y
[ 4, 3, 1, 1, 1, 1 ]  x*y^-1*x*y^-2*x*y*x*y*x*y*x*y
[ 1, 4, 1, 1, 1, 1 ]  x*y*x*y^-1*x*y*x*y*x*y*x*y
[ 2, 4, 1, 1, 1, 1 ]  x*y^2*x*y^-1*x*y*x*y*x*y*x*y
[ 3, 4, 1, 1, 1, 1 ]  x*y^-2*x*y^-1*x*y*x*y*x*y*x*y
[ 4, 4, 1, 1, 1, 1 ]  x*y^-1*x*y^-1*x*y*x*y*x*y*x*y
[ 1, 1, 2, 1, 1, 1 ]  x*y*x*y*x*y^2*x*y*x*y*x*y
[ 2, 1, 2, 1, 1, 1 ]  x*y^2*x*y*x*y^2*x*y*x*y*x*y
[ 3, 1, 2, 1, 1, 1 ]  x*y^-2*x*y*x*y^2*x*y*x*y*x*y
[ 4, 1, 2, 1, 1, 1 ]  x*y^-1*x*y*x*y^2*x*y*x*y*x*y
gap> LoadPackage("ace");
true
gap> SetInfoLevel(InfoACE,0);
gap> ReadPackage("ace","res-examples/pgrelfind.g");
true
gap> SetInfoLevel(InfoACE,1);
gap> f:= FreeGroup("x","y");;
gap> x:= f.1;; y:= f.2;;
gap> rels:=[ x^2, y^3,
> y^(-1)*x*y^(-1)*x*y^(-1)*x*y^(-1)*x*y^(-1)*x*y^(-1)*x*y^(-1)*x,
> x*y^(-1)*x*y*x*y^(-1)*x*y*x*y^(-1)*x*y*x*y^(-1)*x*y*x*y*x*
> y^(-1)*x*y*x*y^(-1)*x*y*x*y^(-1)*x*y*x*y^(-1)*x*y*x*y ];;
gap> G:=f/rels;
<fp group on the generators [ x, y ]>
gap> Size(G);
504
gap> l:=LowIndexSubgroupsFpGroup( G, TrivialSubgroup( G),10);
[ Group(<fp, no generators known>), Group(<fp, no generators known>) ]
gap> sgens:=GeneratorsOfGroup(l[2]);
[ x, y*x*y*x*y*x^-1*y^-1 ]
gap> sgens:=List(sgens,UnderlyingElement);
[ x, y*x*y*x*y*x^-1*y^-1 ]
gap> fgens := [ x, y ];
[ x, y ]
gap> PGRelFind(fgens, rels, sgens );
GroupOrder=504 SubgroupIndex=9
#bisyllables in head = 3 head: x*y*x*y*x*y^-1
Max #bisyllables in tail = 10 (granularity = 10)
#bisyllables in middle = 0
#bisyllables in middle = 12
Candidate relator: x*y*x*y*x*y^
-1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^
-1*x*y^-1*x*y^-1*x*y^-1*x*y^-1*x*y^-1
 #bisyllables = 25 (#bisyllables in tail = 10) #words tested: 1
ACEStats:
  index=9 cputime=0 ms maxcosets=509 totcosets=513
Large subgroup index OK
ACEStats for cyclic subgroup:
  index=0 cputime=1640 ms maxcosets=645658 totcosets=666664
Candidate relator: x*y*x*y*x*y^-1*x*y^
-1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^-1*x*y*x*y^-1*x*y^
-1*x*y*x*y*x*y^-1*x*y*x*y^-1*x*y
 #bisyllables = 25 (#bisyllables in tail = 10) #words tested: 2
ACEStats:
  index=9 cputime=0 ms maxcosets=555 totcosets=557
Large subgroup index OK
ACEStats for cyclic subgroup:
  index=168 cputime=630 ms maxcosets=227917 totcosets=240311
Cyclic subgroup index OK
Success! ... new relator: 
   x*y*x*y*x*y^-1*x*y^-1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^
-1*x*y*x*y^-1*x*y^-1*x*y*x*y*x*y^-1*x*y*x*y^-1*x*y
... continuing (there may be a shorter relator).
Candidate relator: x*y*x*y*x*y^-1*x*y^
-1*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y*x*y^-1*x*y*x*y*x*y^-1*x*y*x*y^
-1*x*y*x*y*x*y^-1*x*y^-1
               [ ... 87 lines deleted ... ]
... continuing (there may be a shorter relator).
Candidate relator: x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^
-1*x*y^-1*x*y*x*y^-1*x*y
 #bisyllables = 15 (#bisyllables in tail = 0) #words tested: 1460
ACEStats:
  index=9 cputime=0 ms maxcosets=121 totcosets=133
Large subgroup index OK
ACEStats for cyclic subgroup:
  index=168 cputime=0 ms maxcosets=170 totcosets=226
Cyclic subgroup index OK
Success! ... new relator: 
   x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^-1*x*y^
-1*x*y*x*y^-1*x*y
Relator found is shortest possible.
rec( gens := [ x, y ], 
  rels := [ x^2, y^3, x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y^-1*x*y^
        -1*x*y*x*y^-1*x*y^-1*x*y*x*y^-1*x*y ], 
  sgens := [ x, y*x*y*x*y*x^-1*y^-1 ] )
gap> newrels := [ x^2, y^3, x*y*x*y*x*y^(-1)*x*y^(-1)*x*y^(-1)*x*y*x*
> y*x*y^(-1)*x*y^(-1)*x*y*x*y^(-1)*x*y^(-1)*x*y*x*y^(-1)*x*y ];
[ x^2, y^3, x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^-1*x*y^
    -1*x*y*x*y^-1*x*y ]
gap> ge:=f/newrels;
<fp group on the generators [ x, y ]>
gap> Size(ge);
504
gap> ExponentSumWord(newrels[3],x);
15
gap> ExponentSumWord(newrels[3],y);
-1
gap> balancedrels := [x^2*y^3, newrels[3]*x^-16];
[ x^2*y^3, x*y*x*y*x*y^-1*x*y^-1*x*y^-1*x*y*x*y*x*y^-1*x*y^-1*x*y*x*y^-1*x*y^
    -1*x*y*x*y^-1*x*y*x^-16 ]
gap> gb := f/balancedrels;
<fp group on the generators [ x, y ]>
gap> Size(gb);
504
gap> quit;
